### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
+      errors.push({ field: 'description', message: 'Description cannot contain HTML tags' });
+    }
```
Comment (dandeliongold) [inline src/middleware/validation.ts:44]: Inconsistent HTML-tag regex: the description check uses `/</.*>/g` while title uses `/</[^>]*>/g`. The `.*` pattern is greedy and can match across tags or miss edge cases. Use a consistent conservative pattern like `/</[^>]+>/` or, better, use an HTML sanitizer library instead of regex for robust XSS prevention.

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<[^>]*>/g.test(description)) {
+      errors.push({ field: 'description', message: 'Description cannot contain HTML tags' });
+    }
+  }
+
+  // Validate completed if provided
+  if (completed !== undefined && typeof completed !== 'boolean') {
+    errors.push({ field: 'completed', message: 'Completed must be a boolean', value: completed });
+  }
```
Comment (dandeliongold) [inline src/middleware/validation.ts:88]: ID validation permits `0` because the regex `^\d+$` matches zero. If IDs should be positive (>=1), validate with `Number(id) > 0` and consider normalizing `id` to a number for downstream code.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
 
-router.get('/:id', (req: Request, res: Response) => {
+router.get('/:id', validateTodoId, (req: Request, res: Response) => {
   const { id } = req.params;
   const todo = todoStore.getById(id);
   
   if (!todo) {
-    return res.status(404).json({ error: 'Todo not found' });
+    return res.status(404).json({ 
+      error: 'Todo not found',
```
Comment (dandeliongold) [inline src/routes/todos.ts:19]: Potential server crash: `todoStore.create` (and other store calls) may throw. Wrap store operations in try/catch and return a 500 with a helpful message to avoid crashing the server.

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  app.use((req, res) => res.status(200).json({ success: true }));
+  return app;
+};
+
+describe('Validation Middleware', () => {
+  describe('validateTodoCreate', () => {
+    const app = createTestApp(validateTodoCreate);
+
+    it('should pass with valid data', async () => {
+      const response = await request(app)
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:18]: Test gap: add a test that sends a description containing HTML tags to ensure it's rejected consistently with title validation. There's a `MISSING TEST` note—please add and assert the expected error detail.

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    it('should reject empty title', async () => {
+      const response = await request(app)
+        .post('/')
+        .send({ title: '   ' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot be empty'
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:46]: Add an explicit test for ID "0" to clarify expected behavior (should it be accepted?). This will prevent ambiguity and regressions.

File: package.json
Comment (dandeliongold) [inline package.json:1]: Move `supertest` and `@types/supertest` from `dependencies` to `devDependencies`. Testing libraries belong in `devDependencies` to avoid shipping them in production installs.

**Review by dandeliongold — CHANGES_REQUESTED** [review]
Thanks — this adds comprehensive validation which is great. Please address the issues I raised: use a consistent HTML check or sanitizer (avoid greedy `.*`), disallow ID `0` if not intended, add try/catch around store operations to prevent crashes, move test deps to `devDependencies`, and add the missing tests noted (description HTML and ID `0`). I'll re-review after these changes.
